src/Transitions/Transition.js

```diff
@@ -0,0 +1,173 @@
+
+function omit(obj, keys) {
+  let included = Object.keys(obj).filter( k => keys.indexOf(k) === -1)
+  let newObj = {};
+
+  included.forEach( key => newObj[key] = obj[key] )
+  return newObj;
+}
+
+function ensureTransitionEnd(node, handler, duration){
+  let fired = false;
+  let done = e => {
+        if (!fired) {
+          fired = true;
+          handler(e);
+        }
+      };
+
+  if ( node ) {
+    TransitionEvents.addEndEventListener(node, done);
+    setTimeout(done, duration);
+  } else {
+    setTimeout(done, 0);
+  }
+}
+
+class Transition extends React.Component {
+
+  constructor(props, context){
+    super(props, context);
+
+    this.state = {};
+
+    this.needsTransition = true;
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.in !== this.props.in) {
+      this.needsTransition = true;
+    }
+  }
+
+  componentDidUpdate() {
+    this.processChild();
+  }
+
+  componentDidMount() {
+    this.processChild();
+  }
+
+  processChild(){
+    let needsTransition = this.needsTransition;
+    let enter = this.props.in;
+
+    if (needsTransition) {
+      this.needsTransition = false;
+      this[enter ? 'performEnter' : 'performLeave']();
+    }
+  }
+
+  performEnter() {
+    let node = React.findDOMNode(this);
+
+    this.props.onEnter(node);
+
+    this.setState({ in: true, transitioning: true }, ()=> {
+      this.props.onEntering(node);
+
+      ensureTransitionEnd(node, () => {
+        if ( this.state.in ){
+          this.setState({
+            transitioning: false
+          }, () => this.props.onEntered(node));
+        }
+
+      }, this.props.duration);
+    });
+  }
+
+  performLeave() {
+    let node = React.findDOMNode(this);
+
+    node = this.props.transitioningNode(node) || node;
+
+    this.props.onExit(node);
+
+    this.setState({ in: false, transitioning: true }, () => {
+      this.props.onExiting(node);
+
+      ensureTransitionEnd(node, () => {
+        if ( !this.state.in ){
+          this.setState({ transitioning: false }, ()=> this.props.onExited(node));
+        }
+      }, this.props.duration);
+    });
+  }
+
+  render() {
+    let childProps = omit(this.props, Object.keys(Transition.propTypes).concat('children'));
+
+    let child = this.props.children;
+    let out = !this.state.in && !this.state.transitioning;
+
+    if ( !child || (this.props.unmountOnExit && out) ){
+      return null;
+    }
+
+    let classes = '';
+
+    // for whatever reason classnames() doesn't actually work here, maybe because they aren't singgle classes?
+    if (this.state.in && !this.state.transitioning) {
+      classes += this.props.enterClassName;
+    }
+
+    if (this.state.in && this.state.transitioning) {
+      classes += this.props.enteringClassName;
+    }
+
+    if (!this.state.in && !this.state.transitioning) {
+      classes += this.props.exitClassName;
+    }
+
+    if (!this.state.in && this.state.transitioning) {
+      classes += this.props.exitingClassName;
```

Comment:
with all this `classes += `
isn't there a possible problem with missing space ?

E.g. `classes = 'class1 class2'`
then
`classes += this.props.nextClassName;` where `nextClassName = 'class3'`
in result:
`classes = 'class1 class2class3'` ?